datalad_next/runners/run.py

```diff
@@ -0,0 +1,258 @@
+
+        def connection_made(self, process: subprocess.Popen) -> None:
+            self.process = process
+            super().connection_made(process)
+
+        def timeout(self, fd: int | None) -> bool:
+            if self.armed and fd is None:
+                self.kill_counter += 1
+                if self.kill_time and self.kill_counter >= self.kill_time:
+                    self.process.kill()
+                    self.kill_time = None
+                if self.terminate_time and self.kill_counter >= self.terminate_time:
+                    self.process.terminate()
+                    self.terminate_time = None
+
+            # If we set the timeout argument due to a not-None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+        def process_exited(self) -> None:
+            self.return_code = self.process.poll()
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+    cmd: list,
+    protocol_class: type[Protocol],
+    *,
+    cwd: Path | None = None,
+    stdin: int | IO | bytes | Queue[bytes | None] | None = None,
+    timeout: float | None = None,
+    closing_action: Callable | None = None,
+    terminate_time: int | None = None,
+    kill_time: int | None = None,
+    **protocol_kwargs,
+) -> Any | Generator:
+    """ A context manager for subprocesses
+
+    The run-context manager will start a subprocess via ``ThreadedRunner``,
+    provide the result of the subprocess invocation, i.e. the result of
+    ``ThreadedRunner.run()`` in the ``as``-variable, and
+    clean up the subprocess when the context is left.
+
+    The run-context manager supports the guaranteed exit of a subprocess through
+    either:
+
+     a) natural exit of the subprocess
+     b) termination of the subprocess via SIGTERM, if ``terminate_time`` is
+        specified
+     c) termination of the subprocess via SIGKILL, if ``kill_time`` is specified
+
+    If the process terminates the run-context manager will ensure that its exit
+    status is read, in order to prevent zombie processes.
+
+    If neither ``terminate_time`` nor ``kill_time`` are specified, and the
+    subprocess does not exit by itself, for example, because it waits for some
+    input, the ``__exit__``-method of the run-context manager will never return.
+    In other words the thread will seem to "hang" when leaving the run-context.
+    The only way to ensure that the context is eventually left is to provide a
+    ``kill_time``. It is a good idea to provide a ``terminate_time`` in
+    addition, to allow the subprocess a graceful exit (see ``kill_time``- and
+    ``terminate_time``-argument descriptions below).
```

Comment:
That makes me wonder: If I have a batch process (no terminate nor kill time set), and my code crashes due to a bug on the receiving end of that batch processes output, the `__exit__` would hang forever?

Wouldn't it make sense to close the process input, and have that trigger a normal exit?

Reply:
I now see that there is `closing_action`. Why would this behavior not be default, and require a user to remember to write this out on each use?